File: src/protocolsession/ProtocolSessionContainer.cpp
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

#include "finalmq/protocolsession/ProtocolSessionContainer.h"
#include "finalmq/streamconnection/StreamConnectionContainer.h"
#include "finalmq/protocolsession/ProtocolRegistry.h"


namespace finalmq {

ProtocolBind::ProtocolBind(hybrid_ptr<IProtocolSessionCallback> callback, const IExecutorPtr& executor, const IExecutorPtr& executorPollerThread, IProtocolFactoryPtr protocolFactory, const std::weak_ptr<IProtocolSessionList>& protocolSessionList, const BindProperties& bindProperties, int contentType)
    : m_callback(callback)
    , m_executor(executor)
    , m_executorPollerThread(executorPollerThread)
    , m_protocolFactory(protocolFactory)
    , m_protocolSessionList(protocolSessionList)
    , m_bindProperties(bindProperties)
    , m_contentType(contentType)
{

}

// IStreamConnectionCallback
hybrid_ptr<IStreamConnectionCallback> ProtocolBind::connected(const IStreamConnectionPtr& connection)
{
    IProtocolPtr protocol;
    IProtocolSessionListPtr protocolSessionList = m_protocolSessionList.lock();
    if (protocolSessionList)
    {
        protocol = m_protocolFactory->createProtocol(m_bindProperties.protocolData);
        assert(protocol);
        IProtocolSessionPrivatePtr protocolSession = std::make_shared<ProtocolSession>(m_callback, m_executor, m_executorPollerThread, protocol, protocolSessionList, m_bindProperties, m_contentType);
        protocolSession->setConnection(connection, !protocol->doesSupportSession());
    }
    return std::weak_ptr<IStreamConnectionCallback>(protocol);
}

void ProtocolBind::disconnected(const IStreamConnectionPtr& /*connection*/)
{
    // should never be called, because the callback will be overriden by connected
    assert(false);
}

bool ProtocolBind::received(const IStreamConnectionPtr& /*connection*/, const SocketPtr& /*socket*/, int /*bytesToRead*/)
{
    // should never be called, because the callback will be overriden by connected
    assert(false);
    return false;
}



//////////////////////////////

ProtocolSessionContainer::ProtocolSessionContainer()
    : m_protocolSessionList(std::make_shared<ProtocolSessionList>())
    , m_streamConnectionContainer(std::make_shared<StreamConnectionContainer>())
{
    m_executorPollerThread = m_streamConnectionContainer->getPollerThreadExecutor();
}

ProtocolSessionContainer::~ProtocolSessionContainer()
{
    terminatePollerLoop();
}

// IProtocolSessionContainer
void ProtocolSessionContainer::init(const IExecutorPtr& executor, int cycleTime, FuncTimer funcTimer, int checkReconnectInterval)
{
    m_executor = executor;
    std::shared_ptr<FuncTimer> pFuncTimer = funcTimer ? std::make_shared<FuncTimer>(std::move(funcTimer)) : nullptr;
    m_streamConnectionContainer->init(cycleTime, [this, pFuncTimer/*, counterInterval*/]() {
        if (pFuncTimer)
        {
            if (m_executor)
            {
                m_executor->addAction([pFuncTimer]() {
                    (*pFuncTimer)();
                });
            }
            else
            {
                (*pFuncTimer)();
            }
        }
        std::vector< IProtocolSessionPrivatePtr > sessions = m_protocolSessionList->getAllSessions();
        for (size_t i = 0; i < sessions.size(); ++i)
        {
            const IProtocolSessionPrivatePtr& session = sessions[i];
            assert(session);
            session->cycleTime();
        }
    }, checkReconnectInterval);
    if (m_executor)
    {
        m_thread = std::thread([this]() { m_streamConnectionContainer->run(); });
    }
}

static std::string endpointToStreamEndpoint(const std::string& endpoint, std::string* protocolName = nullptr)
{
    std::string endpointToUse;
    size_t ixEndOfEndpoint = endpoint.find('{');
    if (ixEndOfEndpoint == std::string::npos)
    {
        endpointToUse = endpoint;
    }
    else
    {
        endpointToUse = endpoint.substr(0, ixEndOfEndpoint);
    }

    size_t ixEndpoint = endpointToUse.find_last_of(':');
    if (ixEndpoint == std::string::npos)
    {
        return "";
    }

    if (protocolName)
    {
        *protocolName = endpointToUse.substr(ixEndpoint + 1, endpointToUse.size() - (ixEndpoint + 1));
    }

    return endpointToUse.substr(0, ixEndpoint);
}


int ProtocolSessionContainer::bind(const std::string& endpoint, hybrid_ptr<IProtocolSessionCallback> callback, const BindProperties& bindProperties, int contentType)
{
    std::string protocolName;
    const std::string endpointStreamConnection = endpointToStreamEndpoint(endpoint, &protocolName);
    if (endpointStreamConnection.empty())
    {
        return -1;
    }

    IProtocolFactoryPtr protocolFactory = ProtocolRegistry::instance().getProtocolFactory(protocolName);
    if (!protocolFactory)
    {
        return -1;
    }

    BindProperties bindPropertiesToUse = bindProperties;
    StreamConnectionContainer::getBindPropertiesFromEndpoint(endpoint, bindPropertiesToUse);

    int err = -1;
    std::unique_lock<std::mutex> lock(m_mutex);
    auto it = m_endpoint2Bind.find(endpoint);
    if (it == m_endpoint2Bind.end())
    {
        ProtocolBindPtr bind = std::make_shared<ProtocolBind>(callback, m_executor, m_executorPollerThread, protocolFactory, m_protocolSessionList, bindPropertiesToUse, contentType);
        m_endpoint2Bind[endpoint] = bind;
        lock.unlock();

        err = m_streamConnectionContainer->bind(endpointStreamConnection, bind, bindPropertiesToUse);
    }
    return err;
}

void ProtocolSessionContainer::unbind(const std::string& endpoint)
{
    std::unique_lock<std::mutex> lock(m_mutex);
    auto it = m_endpoint2Bind.find(endpoint);
    if (it != m_endpoint2Bind.end())
    {
        m_endpoint2Bind.erase(it);
        const std::string endpointStreamConnection = endpointToStreamEndpoint(endpoint);
        m_streamConnectionContainer->unbind(endpointStreamConnection);
    }
}

IProtocolSessionPtr ProtocolSessionContainer::connect(const std::string& endpoint, hybrid_ptr<IProtocolSessionCallback> callback, const ConnectProperties& connectProperties, int contentType)
{
    std::string endpointToUse;
    size_t ixEndOfEndpoint = endpoint.find('{');
    if (ixEndOfEndpoint == std::string::npos)
    {
        endpointToUse = endpoint;
    }
    else
    {
        endpointToUse = endpoint.substr(0, ixEndOfEndpoint);
    }
    size_t ixEndpoint = endpointToUse.find_last_of(':');
    if (ixEndpoint == std::string::npos)
    {
        return nullptr;
    }
    std::string protocolName = endpointToUse.substr(ixEndpoint + 1, endpointToUse.size() - (ixEndpoint + 1));
    IProtocolFactoryPtr protocolFactory = ProtocolRegistry::instance().getProtocolFactory(protocolName);
    if (!protocolFactory)
    {
        return nullptr;
    }

    std::string endpointStreamConnection = endpointToUse.substr(0, ixEndpoint);

    ConnectProperties connectPropertiesToUse = connectProperties;
    StreamConnectionContainer::getConnectPropertiesFromEndpoint(endpoint, connectPropertiesToUse);

    IProtocolSessionPrivatePtr protocolSession = std::make_shared<ProtocolSession>(callback, m_executor, m_executorPollerThread, protocolFactory, m_protocolSessionList, m_streamConnectionContainer, endpointStreamConnection, connectPropertiesToUse, contentType);
    protocolSession->connect();
    return protocolSession;
}


//IProtocolSessionPtr ProtocolSessionContainer::createSession(hybrid_ptr<IProtocolSessionCallback> callback, const IProtocolPtr& protocol, int contentType)
//{
//    assert(protocol);
//    IProtocolSessionPrivatePtr protocolSession = std::make_shared<ProtocolSession>(callback, m_executor, m_executorPollerThread, protocol, m_protocolSessionList, m_streamConnectionContainer, contentType);
//    protocolSession->createConnection();
//    return protocolSession;
//}
//

IProtocolSessionPtr ProtocolSessionContainer::createSession(hybrid_ptr<IProtocolSessionCallback> callback)
{
    IProtocolSessionPrivatePtr protocolSession = std::make_shared<ProtocolSession>(callback, m_executor, m_executorPollerThread, m_protocolSessionList, m_streamConnectionContainer);
    protocolSession->setConnection(nullptr, true);
    return protocolSession;
}



std::vector< IProtocolSessionPtr > ProtocolSessionContainer::getAllSessions() const
{
    std::vector< IProtocolSessionPrivatePtr > protocolSessions = m_protocolSessionList->getAllSessions();
    return { std::make_move_iterator(protocolSessions.begin()), std::make_move_iterator(protocolSessions.end()) };
}

IProtocolSessionPtr ProtocolSessionContainer::getSession(std::int64_t sessionId) const
{
    IProtocolSessionPtr protocolSession = m_protocolSessionList->getSession(sessionId);
    return protocolSession;
}


void ProtocolSessionContainer::run()
{
    if (m_executor == nullptr)
    {
        m_streamConnectionContainer->run();
    }
}


void ProtocolSessionContainer::terminatePollerLoop()
{
    std::vector< IProtocolSessionPtr > sessions = getAllSessions();
    for (size_t i = 0; i < sessions.size(); ++i)
    {
        sessions[i]->disconnect();
    }

    m_streamConnectionContainer->terminatePollerLoop();
    if (m_thread.joinable())
    {
        m_thread.join();
    }
}

IExecutorPtr ProtocolSessionContainer::getExecutor() const
{
    return m_executor ? m_executor : m_executorPollerThread;
}


}   // namespace finalmq

File: csharp/src/protocolsession/ProtocolSessionContainer.cs
//MIT License

//Copyright (c) 2020 bexoft GmbH (mail@bexoft.de)

//Permission is hereby granted, free of charge, to any person obtaining a copy
//of this software and associated documentation files (the "Software"), to deal
//in the Software without restriction, including without limitation the rights
//to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//copies of the Software, and to permit persons to whom the Software is
//furnished to do so, subject to the following conditions:

//The above copyright notice and this permission notice shall be included in all
//copies or substantial portions of the Software.

//THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
//SOFTWARE.

using System.Diagnostics;

namespace finalmq 
{

    public delegate void FuncTimer();

    interface IProtocolSessionContainer : IDisposable
    {
        int CheckReconnectInterval { get; set; }
        void Bind(string endpoint, IProtocolSessionCallback callback, BindProperties? bindProperties = null, int contentType = 0);
        void Unbind(string endpoint);
        IProtocolSession Connect(string endpoint, IProtocolSessionCallback callback, ConnectProperties? connectProperties = null, int contentType = 0);
        IProtocolSession CreateSession(IProtocolSessionCallback callback);
        IList<IProtocolSession> GetAllSessions();
        IProtocolSession GetSession(long sessionId);
        IExecutor? GetExecutor();
    }


    class ProtocolBind : IStreamConnectionCallback
    {
        public ProtocolBind(IProtocolSessionCallback callback, IExecutor? executor, IProtocolFactory protocolFactory, IProtocolSessionList protocolSessionList, BindProperties? bindProperties = null, int contentType = 0)
        {
            m_callback = callback;
            m_executor = executor;
            m_protocolFactory = protocolFactory;
            m_protocolSessionList = protocolSessionList;
            m_bindProperties = bindProperties;
            m_contentType = contentType;
        }

        // IStreamConnectionCallback
        public IStreamConnectionCallback? Connected(IStreamConnection connection)
        {
            IProtocol protocol = m_protocolFactory.CreateProtocol(/* todo m_bindProperties.ProtocolData */);
            IProtocolSessionPrivate protocolSession = new ProtocolSession(m_callback, m_executor, protocol, m_protocolSessionList, m_bindProperties, m_contentType);
            protocolSession.SetConnection(connection, !protocol.DoesSupportSession);
            return protocol;
        }

        public void Disconnected(IStreamConnection connection)
        {
            // should never be called, because the callback will be overriden by connected
            Debug.Assert(false);
        }

        public void Received(IStreamConnection connection, byte[] buffer, int count)
        {
            // should never be called, because the callback will be overriden by connected
            Debug.Assert(false);
        }

        readonly IProtocolSessionCallback m_callback;
        readonly IExecutor? m_executor;
        readonly IProtocolFactory m_protocolFactory;
        readonly IProtocolSessionList m_protocolSessionList;
        readonly BindProperties? m_bindProperties;
        readonly int m_contentType;
    }




    class ProtocolSessionContainer : IProtocolSessionContainer
    {
        public ProtocolSessionContainer(IExecutor? executor)
        {
            m_executor = executor;
        }

        ~ProtocolSessionContainer()
        {
            Dispose(false);
        }

        private void Dispose(bool disposing)
        {
            if (m_disposed)
            {
                return;
            }
            m_disposed = true;

            if (disposing)
            {
                IList<string> endpoints = new List<string>();
                foreach (var entry in m_endpoint2Bind)
                {
                    endpoints.Add(entry.Key);
                }
                foreach (var endpoint in endpoints)
                {
                    Unbind(endpoint);
                }
                m_streamConnectionContainer.Dispose();
            }
        }

        public void Dispose()
        {
            lock (m_mutex)
            {
                Dispose(true);
            }
            GC.SuppressFinalize(this);
        }

        // IProtocolSessionContainer
        public int CheckReconnectInterval 
        {
            get => m_streamConnectionContainer.CheckReconnectInterval;
            set => m_streamConnectionContainer.CheckReconnectInterval = value;
        }
        public void Bind(string endpoint, IProtocolSessionCallback callback, BindProperties? bindProperties = null, int contentType = 0)
        {
            int ixEndpoint = endpoint.LastIndexOf(':');
            if (ixEndpoint == -1)
            {
                throw new ArgumentException("':' missing in endpoint");
            }
            string protocolName = endpoint.Substring(ixEndpoint + 1, endpoint.Length - (ixEndpoint + 1));
            IProtocolFactory protocolFactory = ProtocolRegistry.Instance.GetProtocolFactory(protocolName);

            string endpointStreamConnection = endpoint.Substring(0, ixEndpoint);

            ProtocolBind? bind = null;
            lock (m_mutex)
            {
                if (!m_endpoint2Bind.ContainsKey(endpoint))
                {
                    bind = new ProtocolBind(callback, m_executor, protocolFactory, m_protocolSessionList, bindProperties, contentType);
                    m_endpoint2Bind[endpoint] = bind;
                }
                else
                {
                    throw new ArgumentException("Endpoint already bound: " + endpoint);
                }
            }
            if (bind != null)
            {
                m_streamConnectionContainer.Bind(endpointStreamConnection, bind, bindProperties);
            }
        }
        public void Unbind(string endpoint)
        {
            lock (m_mutex)
            {
                if (m_endpoint2Bind.ContainsKey(endpoint))
                {
                    m_endpoint2Bind.Remove(endpoint);
                    m_streamConnectionContainer.Unbind(endpoint);
                }
            }
        }
        public IProtocolSession Connect(string endpoint, IProtocolSessionCallback callback, ConnectProperties? connectProperties = null, int contentType = 0)
        {
            int ixEndpoint = endpoint.LastIndexOf(':');
            if (ixEndpoint == -1)
            {
                throw new ArgumentException("':' missing in endpoint");
            }
            string protocolName = endpoint.Substring(ixEndpoint + 1, endpoint.Length - (ixEndpoint + 1));
            IProtocolFactory protocolFactory = ProtocolRegistry.Instance.GetProtocolFactory(protocolName);

            IProtocol protocol = protocolFactory.CreateProtocol(/* todo connectProperties.protocolData*/);

            string endpointStreamConnection = endpoint.Substring(0, ixEndpoint);

            IProtocolSessionPrivate protocolSession = new ProtocolSession(callback, m_executor, protocol, m_protocolSessionList, m_streamConnectionContainer, endpointStreamConnection, connectProperties, contentType);
            protocolSession.Connect();
            return protocolSession;
        }
        public IProtocolSession CreateSession(IProtocolSessionCallback callback)
        {
            IProtocolSessionPrivate protocolSession = new ProtocolSession(callback, m_executor, m_protocolSessionList, m_streamConnectionContainer);
            protocolSession.SetConnection(null, true);
            return protocolSession;
        }
        public IList<IProtocolSession> GetAllSessions()
        {
            IList<IProtocolSessionPrivate> protocolSessions = m_protocolSessionList.GetAllSessions();
            IList<IProtocolSession> result = new List<IProtocolSession>(protocolSessions);
            return result;
        }
        public IProtocolSession GetSession(long sessionId)
        {
            IProtocolSession protocolSession = m_protocolSessionList.GetSession(sessionId);
            return protocolSession;
        }
        public IExecutor? GetExecutor()
        {
            return m_executor;
        }

        private readonly IProtocolSessionList m_protocolSessionList = new ProtocolSessionList();
        private readonly IDictionary<string, ProtocolBind> m_endpoint2Bind = new Dictionary<string, ProtocolBind>();
        private readonly IStreamConnectionContainer m_streamConnectionContainer = new StreamConnectionContainer();
        private readonly IExecutor? m_executor = null;
        private bool m_disposed = false;
        private readonly object m_mutex = new object();
    }

}   // namespace finalmq
